mirrord/agent/src/steal/http/hyper_handler/httpv1.rs

```diff
@@ -0,0 +1,194 @@
+            request,
+            self.original_destination,
+            self.handle_version.upgrade_tx.take(),
+            self.filters.clone(),
+            self.port,
+            self.connection_id,
+            self.request_id,
+            self.matched_tx.clone(),
+        ))
+    }
+}
+
+impl HttpV1 {
+    /// Handles the case when no filter matches a header in the request (or we have an HTTP upgrade
+    /// request).
+    ///
+    /// # Flow
+    ///
+    /// 1. Creates a [`http1::Connection`](hyper::client::conn::http1::Connection) to the
+    /// `original_destination`;
+    ///
+    /// 2. Sends the [`Request`] to it, and awaits a [`Response`];
+    ///
+    /// 3. Sends the [`HttpResponse`] back on the connected [`TcpStream`];
+    ///
+    /// ## Special case (HTTP upgrade request)
+    ///
+    /// If the [`Request`] is an HTTP upgrade request, then we send the `original_destination`
+    /// connection, through `upgrade_tx`, to be handled in [`filter_task`](super::filter_task).
+    ///
+    /// - Why not use [`hyper::upgrade::on`]?
+    ///
+    /// [`hyper::upgrade::on`] requires the original [`Request`] as a parameter, due to it having
+    /// the [`OnUpgrade`](hyper::upgrade::OnUpgrade) receiver tucked inside as an
+    /// [`Extensions`](http::extensions::Extensions)
+    /// (you can see this [here](https://docs.rs/hyper/1.0.0-rc.2/src/hyper/upgrade.rs.html#73)).
+    ///
+    /// [`hyper::upgrade::on`] polls this receiver to identify if this an HTTP upgrade request.
+    ///
+    /// The issue for us is that we need to send the [`Request`] to its original destination, with
+    /// [`SendRequest`](hyper::client::conn::http1::SendRequest), which takes ownership of the
+    /// request, prohibiting us to also pass it to the proper hyper upgrade handler.
+    ///
+    /// Trying to copy the [`Request`] is futile, as we can't copy the `OnUpgrade` extension, and if
+    /// we move it from the original `Request` to a copied `Request`, the channel will never
+    /// receive anything due it being in a different `Request` than the one we actually send to
+    /// the hyper machine.
+    // #[tracing::instrument(level = "trace")]
+    async fn unmatched_request(
+        request: Request<Incoming>,
+        upgrade_tx: Option<oneshot::Sender<RawHyperConnection>>,
+        original_destination: SocketAddr,
+    ) -> Result<Response<Full<Bytes>>, HttpTrafficError> {
+        // TODO(alex): We need a "retry" mechanism here for the client handling part, when the
+        // server closes a connection, the client could still be wanting to send a request,
+        // so we need to re-connect and send.
+        let tcp_stream = TcpStream::connect(original_destination)
+            .await
+            .inspect_err(|fail| {
+                error!("Failed connecting to original_destination with {fail:#?}")
+            })?;
+
+        let (mut request_sender, mut connection) = client::conn::http1::handshake(tcp_stream)
```

Comment:
I'll expand on my point about "an enum would have about the same amount of duplication" here.

If we match on the version to do the handshake, then we have 2 separate things being returned, as there is an `http1::Connection` and an `http2::Connection`, so we would need to handle the whole connection thingy inside the each match pattern anyway.

Most of the things in hyper would end up like this, with disparate branches for each HTTP version. So basically the code would look the same, but with `enum` and `match`, instead of having separate `struct`s for each version handling.